thanos compact panic on GetActiveAndPartialBlockIDs

**Thanos, Prometheus and Golang version used**:

```
thanos, version 0.35.0 (branch: HEAD, revision: d9a0efab5709ff6eccc37804996f29990e92ffa5)
  build user:       root@475652f059b4
  build date:       20240502-10:45:11
  go version:       go1.21.9
  platform:         linux/amd64
  tags:             netgo
```

**Object Storage Provider**: S3

**What happened**: It panicked with

```
panic: send on closed channel

goroutine 1416456275 [running]:
github.com/thanos-io/thanos/pkg/block.(*ConcurrentLister).GetActiveAndPartialBlockIDs.func1()
	/app/pkg/block/fetcher.go:260 +0x20c
golang.org/x/sync/errgroup.(*Group).Go.func1()
	/go/pkg/mod/golang.org/x/sync@v0.6.0/errgroup/errgroup.go:78 +0x56
created by golang.org/x/sync/errgroup.(*Group).Go in goroutine 1416456228
	/go/pkg/mod/golang.org/x/sync@v0.6.0/errgroup/errgroup.go:75 +0x96
```

**What you expected to happen**: No panic

**Full logs to relevant components**:

<details>Logs
<p>

```
will update with more info
```

</p>
</details>

**Environment**:
- OS (e.g. from /etc/os-release): official container
- Kernel (e.g. `uname -a`): `Linux thanos-compact-0 5.15.146+ #1 SMP Tue Apr 2 11:55:25 UTC 2024 x86_64 GNU/Linux`

Hi,
I'm seeing the same error both in Store & Compactor. I'm able to reproduce this error on latest thanos 0.36.0 (both store & compactor) as well 

-- On thanos store 
ts=2024-08-09T11:58:11.063840179Z caller=stdlib.go:105 level=debug s3TraceMsg=---------END-HTTP---------

panic: send on closed channel

goroutine 126555170 [running]:
github.com/thanos-io/thanos/pkg/block.(*ConcurrentLister).GetActiveAndPartialBlockIDs.func1()
	/app/pkg/block/fetcher.go:260 +0x20c
golang.org/x/sync/errgroup.(*Group).Go.func1()
	/go/pkg/mod/golang.org/x/sync@v0.7.0/errgroup/errgroup.go:78 +0x56
created by golang.org/x/sync/errgroup.(*Group).Go in goroutine 126555148
	/go/pkg/mod/golang.org/x/sync@v0.7.0/errgroup/errgroup.go:75 +0x96


On Thanos compactor 
===================
caller=runutil.go:117 level=error msg="function failed. Retrying in next tick" err="BaseFetcher: iter bucket: read tcp 100.xx:xxx:xxx:32932->10.xx.xxx.xxxx:80 read: connection reset by peer"
panic: send on closed channel

goroutine 178914629 [running]:
github.com/thanos-io/thanos/pkg/block.(*ConcurrentLister).GetActiveAndPartialBlockIDs.func1()
	/app/pkg/block/fetcher.go:260 +0x20c
golang.org/x/sync/errgroup.(*Group).Go.func1()
	/go/pkg/mod/golang.org/x/sync@v0.7.0/errgroup/errgroup.go:78 +0x56
created by golang.org/x/sync/errgroup.(*Group).Go in goroutine 178914622
	/go/pkg/mod/golang.org/x/sync@v0.7.0/errgroup/errgroup.go:75 +0x96

This issue has to be fixed as its been in around for a while. store & compactor pods have restart in production due to this panic!!

--- a/pkg/block/fetcher.go
+++ b/pkg/block/fetcher.go
@@ -257,7 +257,11 @@ func (f *ConcurrentLister) GetActiveAndPartialBlockIDs(ctx context.Context, ch c
 					mu.Unlock()
 					continue
 				}
-				ch <- uid
+				select {
+				case <-ctx.Done():
+					return ctx.Err()
+				case ch <- uid:
+				}
 			}
 			return nil
 		})

